ext/php7/handlers_exception.c

```diff
@@ -0,0 +1,381 @@
+            }
+        }
+        dd_wrap_exception_or_error_handler(target, is_error_handler);
+    }
+}
+
+static PHP_FUNCTION(ddtrace_set_error_handler) {
+    dd_set_error_handler(INTERNAL_FUNCTION_PARAM_PASSTHRU);
+    dd_set_exception_or_error_handler(&EG(user_error_handler), return_value, true);
+}
+
+static PHP_FUNCTION(ddtrace_set_exception_handler) {
+    dd_set_exception_handler(INTERNAL_FUNCTION_PARAM_PASSTHRU);
+    dd_set_exception_or_error_handler(&EG(user_exception_handler), return_value, false);
+}
+
+static PHP_FUNCTION(ddtrace_restore_exception_handler) {
+    dd_restore_exception_handler(INTERNAL_FUNCTION_PARAM_PASSTHRU);
+
+    if (Z_ISUNDEF_P(&EG(user_exception_handler))) {
+        dd_wrap_exception_or_error_handler(&EG(user_exception_handler), false);
+    }
+}
+
+static zend_internal_function ddtrace_exception_or_error_handler;
+
+ZEND_BEGIN_ARG_INFO_EX(arginfo_ddtrace_exception_or_error_handler, 0, 0, 1)
+ZEND_ARG_INFO(0, exception)
+ZEND_END_ARG_INFO()
+
+#if PHP_VERSION_ID < 70100
+#define ZEND_STR_PREVIOUS "previous"
+#endif
+
+#if PHP_VERSION_ID < 70200
+// zpp API is not safe by itself, but our code is safe here.
+#pragma GCC diagnostic ignored "-Wclobbered"
+#endif
+static PHP_METHOD(DDTrace_ExceptionOrErrorHandler, execute) {
+    volatile zend_bool has_bailout = false;
+    zend_bool is_error_handler = (GC_FLAGS(Z_OBJ_P(getThis())) & GC_IMMUTABLE) != 0;
+    zval *handler = dd_exception_or_error_handler_handler(Z_OBJ_P(getThis()));
+
+    if (is_error_handler) {
+        zend_long type;
+        zend_string *message;
+        zval *error_filename;
+        zend_long error_lineno;
+        zval *symbol_table;
+
+        ZEND_PARSE_PARAMETERS_START(5, 5)
+        Z_PARAM_LONG(type)
+        Z_PARAM_STR(message)
+        Z_PARAM_ZVAL(error_filename)  // may be null
+        Z_PARAM_LONG(error_lineno)
+        Z_PARAM_ZVAL(symbol_table)  // may be null
+        ZEND_PARSE_PARAMETERS_END();
+
+        DDTRACE_G(active_error).type = (int)type;
+        DDTRACE_G(active_error).message = message;
+
+        zval params[5];
+        ZVAL_LONG(&params[0], type);
+        ZVAL_STR(&params[1], message);
+        ZVAL_COPY_VALUE(&params[2], error_filename);
+        ZVAL_LONG(&params[3], error_lineno);
+        ZVAL_COPY_VALUE(&params[4], symbol_table);
+
+        zend_try {
+            // remove ourselves from the stacktrace
+            EG(current_execute_data) = execute_data->prev_execute_data;
+            // this calls into PHP, but without sandbox, as we do not want to interfere with normal operation
+            call_user_function(CG(function_table), NULL, handler, return_value, 5, params);
+        }
+        zend_catch { has_bailout = true; }
+        zend_end_try();
+
+        DDTRACE_G(active_error).type = 0;
+    } else {
+        ddtrace_span_fci *root_span = DDTRACE_G(open_spans_top);
+        zend_object *exception;
+        zval *volatile span_exception;
+        volatile zval old_exception = {0};
```

Comment:
Why are these marked volatile (`has_bailout` as well)?

Reply:
because the compiler must not optimize these into registers due to the zend_try (a setjmp) later on